mono/mini/interp/interp.c

```diff
@@ -6535,42 +6505,40 @@ interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClauseArgs
 			ip ++;
 			MINT_IN_BREAK;
 		MINT_IN_CASE(MINT_ENDFINALLY) {
-			gboolean pending_abort = mono_threads_end_abort_protected_block ();
-			ip ++;
-
-			// After mono_threads_end_abort_protected_block to conserve stack.
-			const int clause_index = *ip;
-
-			// clause_args stores the clause args only for the first frame that
-			// we started executing in interp_exec_method. If we are exiting the
-			// current frame at this finally clause, we need to make sure that
-			// this is the first frame invoked with interp_exec_method.
-			if (clause_args && clause_args->exec_frame == frame && clause_index == clause_args->exit_clause)
-				goto exit_clause;
+			mono_threads_end_abort_protected_block ();
+			guint16 clause_index = *(ip + 1);
 
 			// endfinally empties the stack
 			vt_sp = (guchar*)frame->stack + frame->imethod->total_locals_size;
 			sp = (stackval*)(vt_sp + frame->imethod->vt_stack_size);
 
-			if (finally_ips) {
-				ip = (const guint16*)finally_ips->data;
-				finally_ips = g_slist_remove (finally_ips, ip);
-				/* Throw abort after the last finally block to avoid confusing EH */
-				if (pending_abort && !finally_ips)
-					EXCEPTION_CHECKPOINT;
-				// goto main_loop instead of MINT_IN_DISPATCH helps the compiler and therefore conserves stack.
-				// This is a slow/rare path and conserving stack is preferred over its performance otherwise.
-				goto main_loop;
+			guint16 *ret_ip = *(guint16**)(locals + frame->imethod->clause_data_offsets [clause_index]);
+			if (!ret_ip) {
```

Comment:
Just trying to understand the interpreter a little better here; in the old way, finally_ips was a list and we were just using one element of it. Was it just a stack of finallys we might have to use, and it popped off the top one?

Reply:
Yes. A leave instruction must first invoke all guarding finally clauses associated with the leave. Before, we were using that list as a stack, leave instructions were pushing to it and during each endfinally we were popping one entry. The last entry in the stack would have been the original leave target.